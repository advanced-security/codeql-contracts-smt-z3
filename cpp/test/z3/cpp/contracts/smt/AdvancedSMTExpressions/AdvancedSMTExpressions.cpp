#include "assert.h"
#include <cstddef>

class A {
public:
    int *a;
    int d;
    A *b;
    int operator==(A x)
    {
        return 0;
    }
};

void f1(int i, A* a, A &b){

    /*@ requires @*/
    assert(a);

    /*@ requires @*/
    assert(a->a);

    /*@ requires @*/
    assert(a->d > 0);

    /*@ requires @*/
    assert(a->d);

    /*@ requires @*/
    assert(b.a);
        
    /*@ requires @*/
    assert(a->b->a);

    /*@ requires @*/
    assert(b.b->a);

    /*@ requires @*/
    assert(b.b->b->b->b->b);

    /*@ requires @*/
    assert(a->b->b->b->b->b);
}

