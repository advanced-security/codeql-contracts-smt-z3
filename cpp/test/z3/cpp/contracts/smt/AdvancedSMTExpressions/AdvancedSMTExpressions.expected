| AdvancedSMTExpressions.cpp:15:6:15:7 | specification of f1 | AdvancedSMTExpressions.cpp:18:12:18:12 | a | (not (=  _referenceType__a MaybeNull))\n |
| AdvancedSMTExpressions.cpp:15:6:15:7 | specification of f1 | AdvancedSMTExpressions.cpp:21:15:21:15 | a | (not (=  _referenceType__a_qlcontracts_a MaybeNull))\n |
| AdvancedSMTExpressions.cpp:15:6:15:7 | specification of f1 | AdvancedSMTExpressions.cpp:24:12:24:19 | ... > ... | (>  _valueType__a_qlcontracts_d  0 ) \n |
| AdvancedSMTExpressions.cpp:15:6:15:7 | specification of f1 | AdvancedSMTExpressions.cpp:27:15:27:15 | d | (not (= _valueType__a_qlcontracts_d 0))\n |
| AdvancedSMTExpressions.cpp:15:6:15:7 | specification of f1 | AdvancedSMTExpressions.cpp:30:14:30:14 | a | (not (=  _referenceType__b_qlcontracts_a MaybeNull))\n |
| AdvancedSMTExpressions.cpp:15:6:15:7 | specification of f1 | AdvancedSMTExpressions.cpp:33:18:33:18 | a | (not (=  _referenceType__a_qlcontracts_b_qlcontracts_a MaybeNull))\n |
| AdvancedSMTExpressions.cpp:15:6:15:7 | specification of f1 | AdvancedSMTExpressions.cpp:36:17:36:17 | a | (not (=  _referenceType__b_qlcontracts_b_qlcontracts_a MaybeNull))\n |
| AdvancedSMTExpressions.cpp:15:6:15:7 | specification of f1 | AdvancedSMTExpressions.cpp:39:26:39:26 | b | (not (=  _referenceType__b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b MaybeNull))\n |
| AdvancedSMTExpressions.cpp:15:6:15:7 | specification of f1 | AdvancedSMTExpressions.cpp:42:27:42:27 | b | (not (=  _referenceType__a_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b MaybeNull))\n |
